src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,289 @@
+                    block_ref: &BlockRef::Head,
+                    system_time: Some(Utc::now()),
+                    prev_block_hash: Some(Some(block_hash_to_reset)),
+                    check_object_refs: false,
+                },
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    fn remove_old_files(
+        &self,
+        chain_files_info: &ChainFilesInfo,
+        listener: &Arc<dyn CompactionListener>,
+    ) -> Result<(), CompactError> {
+        listener.begin_phase(CompactionPhase::CleanDatasliceFiles);
+        for file_url in &chain_files_info.data_slice_file_urls {
+            fs::remove_file(file_url.replace("file://", "")).int_err()?;
+        }
+
+        listener.begin_phase(CompactionPhase::CleanBlockFiles);
+        let block_hash_to_reset_string = chain_files_info
+            .block_hash_to_reset
+            .as_multibase()
+            .to_string();
+        for block_file_url in &chain_files_info.block_file_urls {
+            if block_file_url.contains(&block_hash_to_reset_string) {
+                break;
+            }
+            fs::remove_file(block_file_url.replace("file://", "")).int_err()?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl CompactService for CompactServiceImpl {
+    #[tracing::instrument(level = "info", skip_all)]
+    async fn compact_dataset(
+        &self,
+        dataset_handle: &DatasetHandle,
+        dataset_dir_path: &Path,
+        multi_listener: Option<Arc<dyn CompactionMultiListener>>,
+    ) -> Result<(), CompactError> {
+        let listener = multi_listener
+            .and_then(|l| l.begin_compact(dataset_handle))
+            .unwrap_or(Arc::new(NullCompactionListener {}));
+        self.dataset_authorizer
+            .check_action_allowed(dataset_handle, domain::auth::DatasetAction::Read)
+            .await?;
+
+        let dataset = self
+            .dataset_repo
+            .get_dataset(&dataset_handle.as_local_ref())
+            .await?;
+
+        let dataset_kind = dataset
+            .get_summary(GetSummaryOpts::default())
+            .await
+            .int_err()?
+            .kind;
+
+        if dataset_kind != DatasetKind::Root {
+            return Err(CompactError::InvalidDatasetKind(InvalidDatasetKindError {
+                dataset_name: dataset_handle.alias.dataset_name.clone(),
+            }));
+        }
+
+        let compact_dir_path = self.create_compact_dir(dataset_dir_path)?;
+
+        listener.begin_phase(CompactionPhase::GatherChainInfo);
+
+        let chain_files_info = self.get_chain_files_info(dataset.clone()).await?;
+        let new_file_path = self
+            .write_file(
+                chain_files_info.data_slice_file_urls.clone(),
+                &dataset_handle.alias.dataset_name,
+                &compact_dir_path,
+                listener.clone(),
+            )
+            .await?;
+
+        listener.begin_phase(CompactionPhase::ReplaceChainHead);
```

Comment:
Weird name of the phase. I see replacement of data files, deleting old data files. Where is the head replaced?

Reply:
Firstly we reset the head of the dataset chain and commit a new block.